ERCS/erc-post_execution_4337_validation.md

```diff
@@ -0,0 +1,154 @@
+
+Blockchain-based Games
+In-game transactions that depend on specific game states can be validated post-execution. Such validation ensures that actions like trades or character upgrades only proceed if prior operations result in the expected state.
+
+Social Recovery
+Validates the completion of a social recovery process based on approvals from designated guardians, ensuring the integrity of the recovery operation post-execution.
+
+Alternative design considerations
+
+During the design phase, we considered the existing `postOp` function utilized by Paymasters as an alternative mechanism for post-operation validation. As defined within the ERC-4337 standard, Paymaster's approach facilitates gas payment delegation and handles validation immediately after executing each user operation. While functionally significant, this approach cannot inherently assess the state changes produced by the entirety of a user operation bundle. Moreover, the semantics and primary intent of the Paymaster's post-operation validation are focused on the financial aspect of operations.
+
+## Backwards Compatibility
+
+Introducing the `validatePostExecution` method to the IAccount interface and its subsequent implementation in the accounts maintains compatibility with existing account deployments.
+
+Selective Execution
+Similar to the [v0.7.0 selective execution](https://github.com/eth-infinitism/account-abstraction/pull/380#issue-2014371829) approach of utilizing a 4-byte selector within the calldata to signal the EntryPoint contract for post-execution validation, only operations requiring this additional validation step will incur the associated gas costs.
+
+## Security Considerations
+
+### Security Benefits
+Incorporating a post-execution validation mechanism into ERC-4337 brings challenges and security benefits by enabling state validations post-transaction execution that cannot modify state.
+
+#### Enhanced State Integrity Verification
+This mechanism ensures that complex operations involving multiple interdependent transactions can be validated against the final state. It guards against unforeseen changes in state that might occur during the bundle transaction execution, benefiting multi-steps Intents exeucution.
+
+#### State Reentrancy Resistance
+Validating userOps post-bundle execution, the mechanism reduces the risk of reentrancy attacks, which are common in scenarios where multiple operations interact. This is beneficial in complex DeFi interactions, where the final state validation confirms no undesirable state changes have occurred.
+
+### Security Risks
+A concern is ensuring the system's robustness against both Denial of Service (DoS) and griefing attacks, particularly those exacerbated by the potential for increased gas consumption. This section considers mitigation strategies inspired by the original ERC-4337 specification, tailored to address the post-execution validation challenges.
+
+#### Increased Gas Consumption
```

Comment:
yeah, I'll rephrase it, the point is about Sybil, Dos and griefing attacks as described also by the 4337 spec:

> To prevent abuse, we throttle down (or completely ban for a period of time) an entity that causes invalidation of large number of UserOperations in the mempool. To prevent such entities from “sybil-attack”, we require them to stake with the system, and thus make such DoS attack very expensive. Note that this stake is never slashed, and can be withdrawn any time (after unstake delay)